pkg/function/examples/retry/main.go

```diff
@@ -0,0 +1,46 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	functionsdk "github.com/numaproj/numaflow-go/pkg/function"
+	"github.com/numaproj/numaflow-go/pkg/function/server"
+)
+
+var counts map[string]int
+
+const SUCCESS_ITERATION = 3
+
+// This imitates a function which fails the first 2 times it sees a message, but then
+// succeeds on the third time.
+// When it fails, it applies a "retry" Tag to the outgoing message, which causes the message to cycle back to it
+// (this is designed to be used as a Map Vertex which Conditionally Forwards back to
+// itself when the "retry" Tag is set, and forwards onto the next Vertex when it's not set).
+func handle(_ context.Context, keys []string, d functionsdk.Datum) functionsdk.Messages {
+	msgBytes := d.Value()
+	msg := string(msgBytes)
+
+	_, found := counts[msg]
+	if !found {
+		counts[msg] = 1
```

Comment:
This is a bit nit-picking: 1. In a PROD context, we might have multiple messages sharing the same payload value, making the counts incorrect and 2. go map is not thread safe.

But it should be ok considering we are going to use it for E2E tests, which is light weight.

Reply:
1. Yeah, this doesn’t represent a realistic example that anyone would actually use
2. Oh, is this called by multiple goroutines? I didn’t realize
That also reminds me that this would only make sense if there’s a single replica for the Vertex. I will consider if there’s a better example I could have for the e2e test

Reply:
> Is this called by multiple goroutines?

Yes, see https://github.com/numaproj/numaflow/blob/main/pkg/forward/forward.go#L287

Reply:
please use a lock on the map else it will bite the users who blindly copy-paste the code.

Reply:
now using sync.Map